lib/jpm/core.js

```diff
@@ -0,0 +1,235 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const { Cc, Ci } = require("chrome");
+const { OS: {File}} = require("resource://gre/modules/osfile.jsm");
+const { Task } = require("resource://gre/modules/Task.jsm");
+const { Conversion, Status } = require("dev/gcli");
+const { fromFilename: toFileURI } = require("sdk/url");
+const { ZipWriter } = require("./zip");
+const { readManifest } = require("./rdf");
+const { writeBootstrap } = require("./util");
+const { read, remove, isDirectory, exists, list, listTree, uriToPath } = require("./io");
+const { TextDecoder } = require("sdk/io/buffer");
+const { tmpdir } = require("node/os");
+const { uninstall, install, disable, enable } = require("sdk/addon/installer");
+const { set, get } = require("sdk/preferences/service");
+const { AddonManager } = require("resource://gre/modules/AddonManager.jsm");
+const readID = require("jetpack-id/index");
+const path = require("sdk/fs/path");
+
+function getAllAddons() {
+  return new Promise((resolve, reject) => {
+    AddonManager.getAllAddons(resolve);
+  });
+};
+
+const Addon = {
+  name: "Addon",
+  parent: "selection",
+  stringifyProperty: "name",
+  cacheable: true,
+  onInstalled(addon) {
+    this.clearCache();
+  },
+  onUninstalled(addon) {
+    this.clearCache();
+  },
+  constructor() {
+    // Tell GCLI to clear the cache of addons when one is added or removed
+    AddonManager.addAddonListener(this);
+  },
+  lookup() {
+    return getAllAddons().then(addons => {
+      return addons.map(addon => {
+        let name = addon.name + " " + addon.version;
+        name = name.trim().replace(/\s/g, "_");
+        return {name: name,
+                id: addon.id,
+                value: addon};
+      });
+    });
+  }
+};
+exports.Addon = Addon;
+
+const ExistingDirectoryPath = {
+  name: "ExistingDirectoryPath",
+  parent: "string",
+  parse(arg) {
+    const {text:input} = arg;
+    return Task.spawn(function*() {
+      if ((yield exists(input)) &&
+          (yield isDirectory(input)))
+      {
+
+        let predictions = []
+        let entries = yield list(input)
+        for (let entry of entries) {
+          if (yield isDirectory(entry)) {
+            predictions.push({name: entry,
```

Comment:
Why is `incomplete` on a new line? In general this file is very hard to read as it seems limited to like 30 characters, should expand most function arguments out more